### DISCUSSION TITLE
HYPEtools with hourly outputs

### DISCUSSION THREAD

**Discussion by MaartenWynants** [rcapell/HYPEtools#139] [Q&A] [answered]
Hello,

I recently started to run my model with hourly timesteps and I noticed that reading the Xobs and Qobs does not account for these timesteps. I also got some different NSE and KGE values again in plotbasinoutput and HYPE model outputs. 
I was wondering if the HYPEtools already work for hourly timesteps or if this is still a work in progress? 

Regards,
Maarten

**Comment by rcapell** [comment, marked as answer]
Hi Maarten,
hourly timesteps are largely untested (for some functions I did tests with hourly data, but I can't claim they were very thorough...), so anything you observe is worth mentioning here.
I had no hourly Qobs at hand but tested with a Tobs file without problem:
``` R
te <- ReadObs("../testdata/Tobs_hourly.txt", dt.format = "%F %R")
te[, 1]
   [1] "2008-01-01 00:00:00 UTC" "2008-01-01 01:00:00 UTC" "2008-01-01 02:00:00 UTC" "2008-01-01 03:00:00 UTC"
   [5] "2008-01-01 04:00:00 UTC" "2008-01-01 05:00:00 UTC" "2008-01-01 06:00:00 UTC" "2008-01-01 07:00:00 UTC"
timestep(te)
[1] "1hour"
```
Please provide some details on where the functions fail for you!
Best/René

**Reply by MaartenWynants** [reply]
Hej Rene,

I get the following with Xobs and Qobs

>Xobs<-ReadXobs("Xobs.txt")
>Longer-than-daily time steps not allowed in HypeXobs objects.
>Warning message:
>In ReadXobs("Xobs.txt") : Imported as data frame, not as 'HypeXobs' object.
> Xobs[1,1]
>[1] "2016-07-01 UTC"

> Qobs<- ReadPTQobs("Qobs.txt")
> Qobs[1,1]
>[1] "2016-07-01 UTC

> bas <- ReadBasinOutput("results_nocevpam/0000001.txt")
> bas[1,1]
>[1] "2016-07-01 UTC"

The files are imported but the dates are read as dates only, no time. This gives me basically 24 of the same files, it seems to plot every point, but I am not sure what effect it has on the GoF calculations. They are different between the HYPE calculations and the HYPEtools calculations. 
![plot_Q](https://user-images.githubusercontent.com/22454715/193231717-f72c18f5-f149-404d-a52b-e94a0c29aae6.png)
[subass1.txt](https://github.com/rcapell/HYPEtools/files/9682449/subass1.txt)

**Reply by MaartenWynants** [reply]
> timestep(bas)
>[1] "0hour"
> timestep(Xobs)
>NULL
> timestep(Qobs)
>[1] "0hour"

**Reply by rcapell** [reply]
There is no automatic datetime string format recognition, so one has to provide time series formatting as argument `dt.format`, see my example. 
Btw, `ReadPTQobs` is old an old name, better to use `ReadObs` (just a style thing though, `ReadPTQobs` is an alias for the same function call :smile: ).

**Reply by MaartenWynants** [reply]
Aha it imports when specifying the timeformat. 
I think I also found the cause of the diverging GoF values.
<'rNSE' can not be computed: some elements in 'obs' are zero 
I think I should replace the 0 values by NAs.

**Reply by rcapell** [reply]
Good, then I mark it as answered. The general comment still applies, hourly data is poorly tested, so expect bugs along the way :smirk:

**Reply by MaartenWynants** [reply]
As a general question, should I replace the 0 flow values with NA? Since 0 is here a proper observation of no flow. 
This point might be related to my earlier observation that the GoF did not correspond between HYPE and HYPEtools. 
Is it possible that in HYPE, 0 is used in the calculation, while in HYPEtools, it is omitted from the calculation?

Regards,
Maarten

**Reply by rcapell** [reply]
rNSE is not NSE (check the hydroGOF function help). PlotBasinOutput uses function hydroGOF::gof, which computes all measures included in the package and then we pick the ones we want to add to the plot. So the note is not critical unless the function breaks in the process. We could improve the function by replacing the gof call with individual calls. So no, do not replace with NA. If the function breaks or throws a warning, it is a bug (appreciated if you write an issue ticket for that with your basin output file attached for debugging).
